atlassian_jwt_auth/contrib/server/helpers.py

```diff
@@ -32,25 +31,34 @@ def _requires_asap(verifier, auth, parse_jwt_func, build_response_func,
         if verify_issuers_func is not None:
             verify_issuers_func(asap_claims, issuers)
         asap_claim_holder.asap_claims = asap_claims
-    except requests.exceptions.HTTPError as e:
+    except PublicKeyRetrieverException as e:
+        if e.status_code not in (403, 404):
+            # Any error other than "not found" is a problem and should be dealt
+            # with elsewhere.
+            # Note that we treat 403 like 404 to account for the fact that
+            # a server configured to secure directory listings will return 403
+            # for a missing file to avoid leaking information.
+            raise
         # Couldn't find key in key server
         message = 'Unauthorized: Invalid key'
         exception = e
-    except (requests.exceptions.ConnectionError,
-            PublicKeyRetrieverException) as e:
-        message = 'Unauthorized: Backend server connection error'
-        exception = e
     except jwt.exceptions.InvalidIssuerError as e:
-        message = 'Unauthorized: Invalid token issuer'
+        message = 'Forbidden: Invalid token issuer'
```

Comment:
My reading of https://s2sauth.bitbucket.io/spec/ suggests that in all cases where we reject the access token we should be responding with a `401`. So shouldn't we be returning a `401` here?

Reply:
> the access token is accepted but the issuer is not accepted?

Yup: the request has been authenticated, but it doesn't have the right permissions.

The ASAP spec lists RFC 7235 as a reference, but never actually references it. Section 2.1 of RFC 7235 states:

> A server that receives valid credentials that are not adequate to
> gain access ought to respond with the 403 (Forbidden) status code
> (Section 6.5.3 of [RFC7231]).

Section 6.5.3 of RFC 7231 covers 403 responses, about which it says:

> If authentication credentials were provided in the request, the server considers them
> insufficient to grant access.

Reply:
Thank you for clarifying :-) .